### PR TITLE
feat: standardize plugin naming, add getMiddleware fn [WIP]

### PR DESCRIPTION
**Pull request by srchase**
Standardizes naming of plugins.  Use ```resolveConfig``` and ```getMiddleware```.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AllanFly120** [comment]
Can you keep the CHANGELOG untouched? So we have a clue on what they are renamed from.

File: clients/node/client-rds-data-node/RdsDataServiceClient.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      ...configuration
+    });
+    let _config_1 = RegionConfig.resolve(_config_0);
+    let _config_2 = AwsAuthConfig.resolve(_config_1);
+    let _config_3 = EndpointsConfig.resolve(_config_2);
+    let _config_4 = RetryConfig.resolve(_config_3);
+    let _config_5 = UserAgentConfig.resolve(_config_4);
+    super(_config_5);
+    this.config = _config_5;
+    super.use(AwsAuthConfig.getMiddleware(this.config));
```
Comment (AllanFly120) [inline clients/node/client-rds-data-node/RdsDataServiceClient.ts]: I'm not a fan of `getMiddleware()`, because it doesn't give you a middleware. It in-fact mutates the client's middleware stack. Also the all namespace `XXXConfig` doesn't make sense because they are not really `Config`. I still prefer the naming convention of client config component's named `XXXConfig`. If they have middleware to insert, then they are called `XXXPlugin`. If destroy method is required, then the method is called `destroyXXX()`

Comment (srchase) [reply clients/node/client-rds-data-node/RdsDataServiceClient.ts]: I removed the ```Config``` suffix and switched to ```.esolveConfig``` and ```setMiddleware``` to make this more clear.  Using the name for a plugin, along with same function for ```resolveConfig```, ```destroy```, etc will be easier to codegen.

Comment (mtdowling) [reply clients/node/client-rds-data-node/RdsDataServiceClient.ts]: > I'm not a fan of getMiddleware(), because it doesn't give you a middleware. It in-fact mutates the client's middleware stack.

Is this true? Or is this going to be true after some future changes? This code looks like it's getting a list of middleware to inject into `this.use` by calling `getMiddleware`.

> I still prefer the naming convention of client config component's named XXXConfig. If they have middleware to insert, then it should export a XXXPlugin. If destroy method is required, then it should export a method called destroyXXX()

Can you give concrete examples?

Comment (AllanFly120) [reply clients/node/client-rds-data-node/RdsDataServiceClient.ts]: @mtdowling 
>This code looks like it's getting a list of middleware to inject into this.use by calling getMiddleware

I updated the `use` interface recently which takes in a function mutating the middleware stack, see [here](https://github.com/aws/aws-sdk-js-v3/blob/1bb6e3733b792d173fd08f0dcd6993d66b512234/packages/smithy-client/src/client.ts#L22), and [here](https://github.com/aws/aws-sdk-js-v3/blob/1bb6e3733b792d173fd08f0dcd6993d66b512234/packages/signing-middleware/src/middleware.ts#L31-L33). 
A runtime plugin is not just a list of middleware, it may also include mutating existing middleware stack and even delete some middleware if exist. This interface allows more extensibility than only allowing injecting new middleware. 
Using `setMiddleware` here is better than `getMiddleware`. I will propose `getPlugin` as it better describe the function.

> Can you give concrete examples?

* `AwsAuth` customization has input config interface, so `ClientConfiguration` contains [`AwsAuthInput`](https://github.com/srchase/aws-sdk-js-v3/blob/d7c6dad61b290243cd3e542ea81e7b2f2536c493/clients/node/client-rds-data-node/RDSDataConfiguration.ts#L49), and [`AwsAuth.Resolved`](https://github.com/srchase/aws-sdk-js-v3/blob/d7c6dad61b290243cd3e542ea81e7b2f2536c493/clients/node/client-rds-data-node/RDSDataConfiguration.ts#L57). The [`resolveConfig`](https://github.com/srchase/aws-sdk-js-v3/blob/d7c6dad61b290243cd3e542ea81e7b2f2536c493/clients/node/client-rds-data-node/RdsDataServiceClient.ts#L79) should also be called.
* `AwsAuth` customization has middleware stack customization(runtime plugin), so client should [`use the plugin`](https://github.com/srchase/aws-sdk-js-v3/blob/d7c6dad61b290243cd3e542ea81e7b2f2536c493/clients/node/client-rds-data-node/RdsDataServiceClient.ts#L86). But here I suggest renaming it to `super.use(AwsAuth.getPlugin(this.config))`).
* `AwsAuth` doesn't have a `destroy` so take `ProtocolConfig` as an example, because it contains handler. 

```
destroy(): void {
    ClientProtocol.destroy(this.config);
}
```
The `destroy` function will call the destroy function on `this.config.protocol` and hence destroy the handler. The `this.config` is supplied to function because destroyable resources are in the client's resolved config

File: packages/middleware-content-length/src/index.ts [hunk 2]
```diff
@@ -44,11 +44,22 @@ export function contentLengthMiddleware(
...
-  });
-};
+// export const contentLengthPlugin = (options: {
+//   bodyLengthChecker: BodyLengthCalculator;
+// }): Injectable<any, any> => clientStack => {
+//   clientStack.add(contentLengthMiddleware(options.bodyLengthChecker), {
+//     step: "build",
+//     tags: { SET_CONTENT_LENGTH: true }
+//   });
+// };
```
Comment (AllanFly120) [inline packages/middleware-content-length/src/index.ts]: nit: commented code

File: clients/node/client-rds-data-node/RdsDataServiceClient.ts (hunk 1 above)
Comment (mtdowling) [reply clients/node/client-rds-data-node/RdsDataServiceClient.ts]: > I updated the use interface recently which takes in a function mutating the middleware stack, see here, and here.

👍 

> Using setMiddleware here is better than getMiddleware

👍 

> I will propose getPlugin as it better describe the function.

I think `setMiddleware` of `updateMiddleware` makes more sense. It says exactly what the function does. `getPlugin` doesn't convey anything really, and it makes the term "plugin" more ambiguous in the SDK. We need to be really deliberate with how we describe things across codegen and runtime packages.

> AwsAuth customization has input config interface, so ClientConfiguration contains AwsAuthInput, and AwsAuth.Resolved. The resolveConfig should also be called.

Why the difference between `AwsAuthInput` and `AwsAuth.Resolved`? Seems like a strange pattern. If there's some technical limitation in IDEs, then we should standardize on `AwsAuthInput` and `AwsAuthResolved` instead to make this less surprising.

If we do that, then we should get rid of the namespacing altogether and do something like:

1. `AwsAuthConfigInput` interface
2. `AwsAuthConfigResolved` interface
3. `resolveAwsAuthConfig()` function
4. `updateAwsAuthMiddleware()` function

My biggest gripe with the above code is that it's leaning way too much on convention and spreading the responsibility out across so many different disconnected things.

Alternatively we could have a different convention for bridging codegen and runtime code references. For example we could provide a set of interfaces that have generic parameters for the input type, resolved type, and methods for `resolveConfig` and `updateMiddleware`.

File: packages/middleware-content-length/src/index.ts (hunk 2 above)
Comment (srchase) [reply packages/middleware-content-length/src/index.ts]: cleaned up in a6939dd

File: clients/node/client-rds-data-node/RdsDataServiceClient.ts (hunk 1 above)
Comment (AllanFly120) [reply clients/node/client-rds-data-node/RdsDataServiceClient.ts]: > Why the difference between AwsAuthInput and AwsAuth.Resolved? Seems like a strange pattern.

Yes. It is an IDE limitation. It makes sense to get rid of namespace. 

I'm still a little concern about the `updateAwsAuthMiddleware`. It sounds like it takes a parameter that is a middleware to be updated. In fact the [`Injectable` interface](https://github.com/srchase/aws-sdk-js-v3/blob/a6939dda2282f1c22163c35e8594e2faf46cde71/packages/types/src/middleware.ts#L369-L375) is actually the "Plugin" interface we expose(I think we should rename it to `Pluggable`). The `updateAwsAuthMiddleware()` actually a getter of a `Pluggable`, this is why I suggest `getPlugin`. 

> we could provide a set of interfaces that have generic parameters for the input type, resolved type, and methods for resolveConfig and updateMiddleware.

I prefer using convention over interface because by using interface, these config and middleware plugins will be attached to a single object. If you import one of them in client, you import all of them. But for customizations that requires client config and middleware in the commands(there are many), we don't want to import middleware plugin in the client. This will increase the size of the client.

Basically the config and middleware plugin are better to be decoupled.

**Comment by AllanFly120** [comment]
This PR has been replaced to #422
